spyder/plugins/application/plugin.py

```diff
@@ -101,6 +103,33 @@ def on_editor_available(self):
         self.get_container().sig_load_log_file.connect(editor.load)
 
     # -------------------------- PLUGIN TEARDOWN ------------------------------
+    @on_plugin_teardown(plugin=Plugins.Preferences)
+    def on_preferences_teardown(self):
+        preferences = self.get_plugin(Plugins.Preferences)
+        preferences.deregister_plugin_preferences(self)
+
+    @on_plugin_teardown(plugin=Plugins.Editor)
+    def on_editor_teardown(self):
+        editor = self.get_plugin(Plugins.Editor)
+        self.get_container().sig_load_log_file.disconnect(editor.load)
+
+    @on_plugin_teardown(plugin=Plugins.Console)
+    def on_console_teardown(self):
+        if self.is_plugin_available(Plugins.MainMenu):
+            self.report_action.setVisible(False)
+
+    @on_plugin_teardown(plugin=Plugins.Shortcuts)
+    def on_shortcuts_teardown(self):
```

Comment:
Although this is the exact opposite of `on_shortcuts_available`, I think it's not correct.

I say it because deactivating Shortcuts shouldn't remove all other entries added by this plugin to the Help menu. Perhaps the only thing that should be done here is to remove the shortcuts action from the app context menu (see `get_application_context_menu`).